teps/0109-better-structured-provenance-retrieval-in-tekton-chains.md

```diff
@@ -0,0 +1,478 @@
+
+The workflow for the targeted results and params in Chains is like the following:
+```
+(TaskRun params & results) —-> (Scanned by Chains controller, targets are collected) —> (Targets are parsed and validated and matched against with defined result and param structures)  —-> (Chains controller sign the signable targets and put into intoto attestations)
+```
+As mentioned above, the signable targets can be split into inputs (sources) and outputs (artifacts). Instead of scatteredly collecting provenance pieces from TaskRun, we collect inputs/ provenance from params, and generate outputs provenance from TaskRun results. The signable targets being found in params will be the inputs, and those found in TaskRun results will be the outputs.
+
+Tekton Chains will follow [TEP 0075](https://github.com/tektoncd/community/blob/main/teps/0075-object-param-and-result-types.md) and [TEP 0076](https://github.com/tektoncd/community/blob/main/teps/0076-array-result-types.md) and support the sets of json schemas and described in the documents. In this proposal, the expected formats of params and results are also defined.
+
+
+### Notes and Caveats
+
+When these well defined structures grow as richer provenance are pursued, the size of the structure can eventually grow beyong the limit of the container termination message. [TEP 0086](https://github.com/tektoncd/community/pull/521) is addressing this issue.
+
+
+## Design Details
+
+### Signable Targets
+
+All signable targets should be provided in a structure that is well defined in the proposal. If any field is missed, the signable targets will be skipped and an error will be thrown.
+
+#### Version Control Sources
+
+Based on the use cases and data storage within provenance described above, it makes sense to have the source object to be the following because all version control sources will be grinded down into the requirements in the Predicate’s materials:
+```
+  ...
+  - name: CHAINS-VCS
+    type: object
+    description: |
+      The source distribution
+      * uri: resource uri of the git commit
+      * digest: revision digest
+      * alg: digest algorithm: sha256 etc.
+      * src_type: source VCS type: git, hg, etc.
+    properties:
+    src_type:
+      type: string
+    uri:
+      type: string
+    alg:
+      type: string
+    digest:
+      type: string
+```
+The required formats for src_type is one of `git`, `hg`, `svc`, `baz`, `cvs`, `perforce`. The source types can be easily extended when more VCS are supported. The required formats for uri should follow [ResourceURI](https://github.com/in-toto/attestation/blob/main/spec/field_types.md#field-type-definitions). The required format for alg is one of pre-defined algorithm names in intoto attestation [field type definitions](https://github.com/in-toto/attestation/blob/main/spec/field_types.md#field-type-definitions). The digest is the revision digest.
```

Comment:
Is there a reason why we need to constrain the source type? 👀

Reply:
Ah maybe my wording is hinting this, but our source types are not constrained. This can be extended later. I changed my wording to make it clearer